File: cpp/allergies/allergies.cpp
#include "allergies.h"

namespace allergies {
	   
	allergy_test::allergy_test(int score) : _score(score) {

		for (const auto& allergen : allergenValues) {
			if (allergen.second & score)
				_allergies.insert(allergen.first);
		}
	};
	
	bool allergy_test::is_allergic_to(std::string allergen) const
	{
		return allergenValues.at(allergen) & _score;
	}

	std::unordered_set<std::string> allergy_test::get_allergies() const
	{
		return _allergies;
	}

}  // namespace allergies

File: java/allergies/src/main/java/Allergies.java
import java.util.List;
import java.util.stream.Collectors;

class Allergies {

    private final int score;

    Allergies(int score) {
        this.score = score;
    }

    public boolean isAllergicTo(Allergen allergen) {
        // Allergens are signified by a bit flag in an 8-bit int.
        return (score & allergen.getScore()) > 0;
    }

    public List<Allergen> getList() {
        return Allergen
                .stream()
                .filter(this::isAllergicTo)
                .collect(Collectors.toList());
    }
}
